erigon-lib/recsplit/multiencseq/sequence_reader.go

```diff
@@ -0,0 +1,170 @@
+	// plain elias fano (legacy)
+	if data[0]&0b10000000 == 0 {
+		return eliasfano32.Count(data)
+	}
+
+	// rebased elias fano
+	if data[0] == 0x90 {
+		return eliasfano32.Count(data[1:])
+	}
+
+	// simple encoding
+	if data[0]&0b11110000 == 0b10000000 {
+		return uint64(data[0]&0b00001111) + 1
+	}
+
+	panic("unknown encoding")
+}
+
+// TODO: optimize me - to avoid object allocation (this TODO was inherited from elias_fano.go)
+func Seek(baseNum uint64, data []byte, n uint64) (uint64, bool) {
+	seq := ReadMultiEncSeq(baseNum, data)
+	return seq.search(n)
+}
+
+func (s *SequenceReader) Get(i uint64) uint64 {
+	if s.currentEnc == SimpleEncoding {
+		return s.sseq.Get(i)
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		return s.ref.Get(i)
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) Min() uint64 {
+	if s.currentEnc == SimpleEncoding {
+		return s.sseq.Min()
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		return s.ref.Min()
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) Max() uint64 {
+	if s.currentEnc == SimpleEncoding {
+		return s.sseq.Max()
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		return s.ref.Max()
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) Count() uint64 {
+	if s.currentEnc == SimpleEncoding {
+		return s.sseq.Count()
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		return s.ref.Count()
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) Reset(baseNum uint64, raw []byte) {
+	// plain elias fano (legacy)
+	if raw[0]&0b10000000 == 0 {
+		s.currentEnc = PlainEliasFano
+		s.ref.Reset(0, raw)
+		return
+	}
+
+	// rebased elias fano
+	if raw[0] == 0x90 {
+		s.currentEnc = RebasedEliasFano
+		s.ref.Reset(baseNum, raw[1:])
+		return
+	}
+
+	// simple encoding
+	if raw[0]&0b11110000 == 0b10000000 {
+		s.currentEnc = SimpleEncoding
+		s.sseq.Reset(baseNum, raw[1:])
+		return
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) search(v uint64) (uint64, bool) {
+	if s.currentEnc == SimpleEncoding {
+		return s.sseq.Search(v)
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		return s.ref.Search(v)
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) Iterator(v int) stream.U64 {
+	if s.currentEnc == SimpleEncoding {
+		it := s.sseq.Iterator()
+		if v > 0 {
+			it.Seek(uint64(v))
+		}
+		return it
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
+		it := s.ref.Iterator()
+		if v > 0 {
+			it.Seek(uint64(v))
+		}
+		return it
+	}
+
+	panic("unknown encoding")
+}
+
+func (s *SequenceReader) ReverseIterator(v int) stream.U64 {
+	if s.currentEnc == SimpleEncoding {
+		it := s.sseq.ReverseIterator()
+		if v > 0 {
+			it.Seek(uint64(v))
+		}
+		return it
+	} else if s.currentEnc == PlainEliasFano || s.currentEnc == RebasedEliasFano {
```

Comment:
btw since constants are growing could just do s.currentEnc < threshold. So maybe reorder them as PlainEF, RebasedEF, and then SimpleEnc?

Reply:
hm, problem here is that the consts are at the same time the type itself and the bit fragments that are used when encoding them, i.e., they are not 1, 2 and 3, but 0, 128 and 144.

I think I could do it, but I think the code now is clearer to read and understand?

Also, if we are going to merge it now, maybe when can even remove the plain elias-fano choice since it would not be used anymore in this part of code and have one less comparison.